### ISSUE TITLE
On Demand Runners for GitHub Actions

### ISSUE THREAD

**Issue by fkorotkov** [cirruslabs/tart#237]
GitHub has a [way for auto-scaling runners for Actions](https://docs.github.com/en/actions/hosting-your-own-runners/autoscaling-with-self-hosted-runners#about-autoscaling) where one subscribes to `workflow_job` events launches runners on demand.

We can potentially develop such service as a paid offering. The service will look like a proprietary GitHub App that handles the events and some kind of daemon on Mac hosts that connects to the GitHub App and spins up Tart VMs on demand based on `workflow_job` events.

GitHub Actions workflow configuration will look like:

```yaml
runs-on: [self-hosted, cirrus, ghcr.io/cirruslabs/macos-ventura-xcode:latest]
```

`cirrus` label will indicate to use our managed runners and the last label will indicate which Tart VM to use.

**Comment by wegotoeleven** [comment]
This is _awesome_

**Comment by naikrovek** [comment]
The GitHub app will send webhooks somewhere and that "somewhere" can take action based on them and their contents.  

I wrote something very similar at work and am planning on making the tool I have which wraps `tart` do this as well.  I like when I see others have similar ideas because it means those ideas that I've had probably aren't completely terrible.

**Comment by edude03** [comment]
I've built something similar as well

**Comment by fkorotkov** [comment]
[We've tried to reach out to GitHub folks](https://github.com/actions/runner-images/issues/2187#issuecomment-1122350871) to collaborate on using Tart to power Apple Silicon runners for Actions but haven't got a response back...

So we've built our own service with on-demand runners for actions powered by Tart. Documenting it in #288. Feel free to try it out via installing this app https://github.com/apps/cirrus-runners for your organization.

**Comment by naikrovek** [comment]
github are often slow to respond.  I have contacts internally that I can reach out to, if you like.

I'm using tart to prototype macos runners for my enterprise.  I'll describe it to them when I meet with them next.

**Comment by fkorotkov** [comment]
@naikrovek yeah, I've also reached out thought many channels with no luck. Appreciate if you'll mention us.

Note that even for the Very Serious Mac Pro you'll be able to run only two VMs a host. In our case for Cirrus Runners we have a lot of Mac Minis in two different data centers. The app orchestrates them, does health monitoring of hosts, uses GitHub API to monitor health from the GitHub side, etc. We've re-purposed a lot of what we do with Cirrus CI for years. 😎

**Comment by naikrovek** [comment]
Why only two VMs per host?  Is that a Virtualization.Framework limitation or something else?

**Comment by fkorotkov** [comment]
That's a limitation from macOS Terms of Service and it's also enforced by the `Virtualization.Framework`.

**Comment by naikrovek** [comment]
well I can buy additional copies of MacOS to increase VM entitlements, I thought.  Shame that it's enforced in the framework, given that you can buy the right to have more...  :(
